api/envoy/extensions/retry/admission_control/concurrency_budget/v3/concurrency_budget_config.proto

```diff
@@ -0,0 +1,52 @@
+syntax = "proto3";
+
+package envoy.extensions.retry.admission_control.concurrency_budget.v3;
+
+import "envoy/type/v3/percent.proto";
+
+import "google/protobuf/wrappers.proto";
+
+import "udpa/annotations/status.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.retry.admission_control.concurrency_budget.v3";
+option java_outer_classname = "ConcurrencyBudgetConfigProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/extensions/retry/admission_control/concurrency_budget/v3;concurrency_budgetv3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Concurrency Budget Retry Admission Control]
+
+// The concurrency budget retry admission controller allows for limiting the number of concurrent
+// retries that can be in flight at any given time to the same upstream cluster.
+// This is useful for preventing a single upstream cluster from being overloaded with retries.
+//
+// The concurrency budget limits the number of concurrent retries as a percentage of the total
+// number of active tries (retries or otherwise) to the cluster at the time the retry enters backoff.
+//
+// Retries are counted as active as soon as they enter backoff,
+// until they are either successful (response streaming starts) or aborted.
+//
+// Tries meanwhile include both the initial try, and all retries. The initial try is counted as
+// active as soon as it is dispatched, and remains active until the upstream stream is ended.
+//
+// [#extension: envoy.retry_admission_control.concurrency_budget]
+// [#not-implemented-hide:]
+message ConcurrencyBudgetConfig {
```

Comment:
I thought we were going with the sliding window approach for this one? See https://github.com/envoyproxy/envoy/issues/30205#issuecomment-1811044833. If you intentionally did it this way can you explain why it's preferred?

Reply:
My goal to start with was to move the existing circuit breakers into extensions. Once we've gotten that done, then we can work on adding the new feature as a new extension. If you think it'd be helpful I can draft up a proto config for what I'm intending to call that the `RateBudget` limiter.

The goal was to add features without impacting existing behavior so that people's existing retry budget setups don't get messed up things changed so much.

Reply:
> The goal was to add features without impacting existing behavior so that people's existing retry budget setups don't get messed up things changed so much.

Fair, but this is why we have the deprecation policy and runtime guards. Folks will have ~1 year to use the new configuration and should be able to replicate today's retry budget behavior if the lookback window size is zero. You wouldn't be looking at a time range in that case, but instead the instantaneous number in-flight retries.

Reply:
> if the lookback window size is zero. You wouldn't be looking at a time range in that case, but instead the instantaneous number in-flight retries.

Is the proposal here that the extension acts differently if the window size is 0? I think you'd need to change a lot of the implementation to make it act as a concurrency-based limit vs a rate-based limit, at least the way I'm thinking about it.

If the implementation of the two is significantly different, I guess I don't see the benefit to keeping them in one extension vs splitting it up. We could choose to either immediately deprecate or just hide the concurrency-based implementation to discourage it.

Reply:
My mental model for the sliding window approach isn't quite a rate limiter. I imagined we'd have something like:

* Define a budget window, `tw`.
* Count up the number of retries and non-retry requests fired off in the time range `[now-tw, now]`.
* If `num_retries / (num_retries + num_requests)` >= budget, don't retry.

One way to think about it is we increment `num_requests` when sending a request, but only decrement `num_requests` when time `tw` has passed after getting the reply (or whatever the result is). If that's how it works, when `tw` is 0 we are only really counting active requests/retries like we do in today's retry budget implementation. As we increase `tw`, it starts to look like a sliding window enforcement.

Does that make sense? You've thought about this more than I have recently, so call out anything about this that you think looks fishy.

Reply:
> If the implementation of the two is significantly different, I guess I don't see the benefit to keeping them in one extension vs splitting it up. We could choose to either immediately deprecate or just hide the concurrency-based implementation to discourage it.

If it turns out that my mental model in the last comment is totally wrong, then you're absolutely right. It would make no sense to keep them in the same extension. Just let me know what you think when you have a chance.

Reply:
So the way I was thinking about it is that the rate limiter is using a token bucket approach to count when tries to the upstream _start_. Each upstream try starting provides a token (or part of a token), where each token allows a retry to be issued. This implementation actually doesn't care at all about when a try ends, only when one begins. In the same vein, we don't actually know how many active tries there are at any given point.

On the other hand with the concurrency budget, we need to track both when a try starts and when it ends. We need to know how many requests are active _right now_ to know how many retries can be allowed. This doesn't need the token bucket approach at all, we can just keep a gauge that tracks how many active retries and total tries there are at the current time.

Anyways, yeah I don't think there's a ton of common implementation space and it'd probably be better to keep them separate.

By the way I'm also happy to call/slack/etc if you'd prefer that, but async is also good if that works.

Reply:
Yeah, a synchronous conversation would be useful here. Just ping me on slack (tonya11en) and we can figure out a good time to have a call.